Std/Lean/Meta/Iterator.lean

```diff
@@ -0,0 +1,68 @@
+import Lean.Meta.Basic
+
+namespace Lean.Meta
+
+/--
+Provides an iterface for iterating over values that are defined in a particular
+meta state.
+-/
+structure Iterator (α : Type) where
+  /-- Function for getting next value and state pair. -/
+  next : MetaM (Option (α × Meta.SavedState))
```

Comment:
I imagine we might want to define a more general notion of `Iterator` at some point in the future. Should we call this one `MetaIterator`, or make it `protected`, to prevent future name-clash problems?

Reply:
I agree.  I explored going the more general route directly to a polymorphic `BacktrackingIterator`, but it turns out to need several classes, and I think is a premature generalization at the moment.

So I just made this one protected.